ARROW-9618: [Rust] [DataFusion] Made it easier to write optimizers

jorgecarleitao opened the pull request:
This PR adds 5 auxiliary functions to the optimizer package that allow us to more easily write optimizers and re-writes the optimizers using them.

This PR assumes that the majority of the optimizations will:

* recurse logical expressions and re-write them (currently we have the type coercion)
* recurse logical plans and re-write them (currently we have the projection pushdown)

To transverse and re-write expressions, this PR introduces two functions:

```
/// Returns all expressions composing the expression.
/// E.g. if the expression is "(a + 1) + 1", it returns ["a + 1", "1"] (as Expr objects)
fn expr_expressions(expr: &Expr) -> Result<Vec<&Expr>>

/// returns a new expression where the expressions in expr are replaced by the ones in `expr`.
/// This is used in conjunction with ``expr_expressions`` to re-write expressions.
pub fn from_expression(expr: &Expr, expressions: &Vec<Expr>) -> Result<Expr> {
```

these are used on an optimizer as follows:

```
let mut expressions = expr_expressions(expr)?;

# recurse on the expression
let mut expressions = expressions.iter().map(|e| recursion(e)).collect()

# modify `expressions` (optimizer's code)

utils::from_expression(expr, &expressions)
```

Likewise, this PR introduces 3 functions for the plans:

```
/// returns all expressions in the logical plan.
pub fn expressions(plan: &LogicalPlan) -> Vec<Expr>

/// returns all inputs in the logical plan
pub fn inputs(plan: &LogicalPlan) -> Vec<&LogicalPlan>

/// Returns a new logical plan based on the original one with inputs and expressions replaced
pub fn from_plan(
    plan: &LogicalPlan,
    expr: &Vec<Expr>,
    inputs: &Vec<LogicalPlan>,
) -> Result<LogicalPlan>
```

Overall, these make writing optimizers a much developer-friendlier experience because expressions or plans that the optimizer does not care can just be left alone.

andygrove commented:
@jorgecarleitao Is this still a draft or is it ready for review now?

jorgecarleitao commented:
Ups, forgot to change the status. Ready to review, @andygrove .

andygrove commented:
I did a quick review of this PR and it looks good I think. @jorgecarleitao would you mind rebasing it and I'll find time this week for a more thorough review.

jorgecarleitao commented:
Thanks, @andygrove . This is now rebased.

jorgecarleitao commented:
@alamb , I agree with you that we will need to migrate to a visitor pattern. The goal of this PR is to write functions that walk and re-write the tree, so that we do not have to copy-paste stuff when writing an optimizer. This will make it easier to migrate, since there is well defined pattern in place to migrate from.

@andygrove and @alamb , I rebased this against master again.

alamb commented:
This looks good to me

jorgecarleitao commented:
Incorporated [this comment](https://github.com/apache/arrow/pull/7880#discussion_r470753855) from @alamb and squashed the whole thing in two commits. Thanks @alamb for pushing for the assert there. It makes total sense. 👍
